File: wheystore/wheystore.Web/Modules/SanPhamDB/SanPham/SanPhamForm.cs
using wheystore.SanPhamDB;
using System.Collections.Generic;


namespace wheystore.SanPhamDB.Forms;

[FormScript("SanPhamDB.SanPham")]
[BasedOnRow(typeof(SanPhamRow), CheckNames = true)]
public class SanPhamForm
{
    public string TenSp { get; set; }
    public decimal GiaBan { get; set; }
    public int SoLuong { get; set; }
    public string MoTa { get; set; }
    [DisplayName("Category")]
    [LookupEditor(typeof(DanhMucSanPhamRow))]
    public int MaLoai { get; set; }

    [DisplayName("Supplier")]
    [ServiceLookupEditor(typeof(NhaCungCapRow))]
    public int NhaCungCap { get; set; }

    [DisplayName("Thương Hiệu")]
    [LookupEditor(typeof(ThuongHieuRow))]
    public int ThuongHieuId { get; set; }
    public string AnhSp { get; set; }
    public string AnhSpCt { get; set; }

    // Giảm giá % từ 0 -> 100, không cho nhập số lẻ
    [DecimalEditor(MinValue = 0, MaxValue = 100, Decimals = 0)]
    public decimal GiamGia { get; set; }
    // Giá khuyến mãi: để readonly, chỉ cho code set
    [ReadOnly(true)]

    
    public decimal GiaKhuyenMai { get; set; }
    public bool HienThi { get; set; }

    

    public List<int> HuongViList { get; set; }

    public DateTime CreatedAt { get; set; }
    public DateTime UpdatedAt { get; set; }


}

File: wheystore/wheystore.Web/Modules/ServerTypes/SanPhamDB/SanPhamForm.ts
import { BooleanEditor, DateEditor, DecimalEditor, initFormType, IntegerEditor, LookupEditor, PrefixedContext, ServiceLookupEditor, StringEditor } from "@serenity-is/corelib";

export interface SanPhamForm {
    TenSp: StringEditor;
    GiaBan: DecimalEditor;
    SoLuong: IntegerEditor;
    MoTa: StringEditor;
    MaLoai: LookupEditor;
    NhaCungCap: ServiceLookupEditor;
    ThuongHieuId: LookupEditor;
    AnhSp: StringEditor;
    AnhSpCt: StringEditor;
    GiamGia: DecimalEditor;
    GiaKhuyenMai: DecimalEditor;
    HienThi: BooleanEditor;
    HuongViList: LookupEditor;
    CreatedAt: DateEditor;
    UpdatedAt: DateEditor;
}

export class SanPhamForm extends PrefixedContext {
    static readonly formKey = 'SanPhamDB.SanPham';
    private static init: boolean;

    constructor(prefix: string) {
        super(prefix);

        if (!SanPhamForm.init) {
            SanPhamForm.init = true;

            var w0 = StringEditor;
            var w1 = DecimalEditor;
            var w2 = IntegerEditor;
            var w3 = LookupEditor;
            var w4 = ServiceLookupEditor;
            var w5 = BooleanEditor;
            var w6 = DateEditor;

            initFormType(SanPhamForm, [
                'TenSp', w0,
                'GiaBan', w1,
                'SoLuong', w2,
                'MoTa', w0,
                'MaLoai', w3,
                'NhaCungCap', w4,
                'ThuongHieuId', w3,
                'AnhSp', w0,
                'AnhSpCt', w0,
                'GiamGia', w1,
                'GiaKhuyenMai', w1,
                'HienThi', w5,
                'HuongViList', w3,
                'CreatedAt', w6,
                'UpdatedAt', w6
            ]);
        }
    }
}